File: backend/src/Caro.Core.Domain/Entities/Player.cs
namespace Caro.Core.Domain.Entities;

/// <summary>
/// Represents a player in the Caro game.
/// None represents an empty cell or no current player.
/// </summary>
public enum Player
{
    /// <summary>No player (empty cell)</summary>
    None,

    /// <summary>Red player (moves first)</summary>
    Red,

    /// <summary>Blue player (moves second)</summary>
    Blue
}

/// <summary>
/// Extension methods for Player enum
/// </summary>
public static class PlayerExtensions
{
    /// <summary>
    /// Get the opponent of the given player
    /// </summary>
    public static Player Opponent(this Player player) => player switch
    {
        Player.Red => Player.Blue,
        Player.Blue => Player.Red,
        _ => Player.None
    };

    /// <summary>
    /// Check if the player is valid (not None)
    /// </summary>
    public static bool IsValid(this Player player) => player is Player.Red or Player.Blue;

    /// <summary>
    /// Check if the player is Red
    /// </summary>
    public static bool IsRed(this Player player) => player == Player.Red;

    /// <summary>
    /// Check if the player is Blue
    /// </summary>
    public static bool IsBlue(this Player player) => player == Player.Blue;
}

File: backend/internal/domain/player.go
package domain

type Player int

const (
	PlayerNone Player = iota
	PlayerRed
	PlayerBlue
)

func (p Player) Opponent() Player {
	switch p {
	case PlayerRed:
		return PlayerBlue
	case PlayerBlue:
		return PlayerRed
	default:
		return PlayerNone
	}
}

func (p Player) IsValid() bool {
	return p == PlayerRed || p == PlayerBlue
}

func (p Player) String() string {
	switch p {
	case PlayerRed:
		return "red"
	case PlayerBlue:
		return "blue"
	default:
		return "none"
	}
}

func ParsePlayer(s string) (Player, bool) {
	switch s {
	case "red":
		return PlayerRed, true
	case "blue":
		return PlayerBlue, true
	case "none":
		return PlayerNone, true
	default:
		return PlayerNone, false
	}
}
